torch/csrc/inductor/aot_runtime/model.h

```diff
@@ -40,9 +41,101 @@
 namespace torch {
 namespace aot_inductor {
 
-using RAIIAtenTensorHandle = std::shared_ptr<AtenTensorOpaque>;
+// UniqueAtenTensorHandle is used to manage input, output, and intermediate
+// buffers
+class UniqueAtenTensorHandle {
+ public:
+  UniqueAtenTensorHandle() = delete;
+  UniqueAtenTensorHandle(const UniqueAtenTensorHandle& other) = delete;
+  UniqueAtenTensorHandle& operator=(const UniqueAtenTensorHandle& other) =
+      delete;
+
+  // Steal the ownership from another UniqueAtenTensorHandle using std::move
+  UniqueAtenTensorHandle(UniqueAtenTensorHandle&& other) = default;
+  UniqueAtenTensorHandle& operator=(UniqueAtenTensorHandle&& other) = default;
+
+  // Steal the ownership from raw AtenTensorHandle
+  UniqueAtenTensorHandle(AtenTensorHandle handle)
+      : handle_(handle, [](AtenTensorHandle ptr) {
+          AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_delete_tensor_object(ptr));
+        }) {}
+
+  ~UniqueAtenTensorHandle() {
+    handle_.reset();
+  }
 
-using ConstantMap = std::unordered_map<std::string, RAIIAtenTensorHandle>;
+  // Return a raw AtenTensorHandle to be used by aoti_torch functions
+  // Note: this function does NOT transfer the ownership of the handle
+  operator AtenTensorHandle() const {
+    return handle_.get();
+  }
+
+  AtenTensorHandle release() {
+    return handle_.release();
+  }
+
+  void reset() {
+    handle_.reset();
+  }
+
+ private:
+  std::unique_ptr<AtenTensorOpaque, std::function<void(AtenTensorOpaque*)>>
+      handle_;
+};
+
+// Constant weights are shared among model instances, and thus the ownership
+// needs to be shared
```

Comment:
Can you tell me why you even need this? It seems to me that the weights would never get deallocated so you don't need to refcount them from inside the model code at all??

Reply:
I was thinking we need a wrapper class to provide a .get() API, but since I have removed that in the codegen, I agree we don't need this. Changing.

Reply:
Even if you do need the .get() API, you can just have a wrapper class that holds a raw pointer and provides get() to it.